stan/math/prim/mat/fun/ocl_gpu.hpp

```diff
@@ -0,0 +1,292 @@
+  kernel_groups["add"] = "basic_matrix";
+  kernel_groups["subtract"] = "basic_matrix";
+  kernel_groups["copy_submatrix"] = "basic_matrix";
+  kernel_groups["scalar_mul_diagonal"] = "matrix_multiply";
+  kernel_groups["scalar_mul"] = "matrix_multiply";
+  kernel_groups["basic_multiply"] = "matrix_multiply";
+  kernel_groups["multiply_self_transposed"] = "matrix_multiply";
+  kernel_groups["lower_tri_inv_step1"] = "matrix_inverse";
+  kernel_groups["lower_tri_inv_step2"] = "matrix_inverse";
+  kernel_groups["lower_tri_inv_step3"] = "matrix_inverse";
+  kernel_groups["cholesky_block"] = "cholesky_decomposition";
+  kernel_groups["check_nan"] = "check_gpu";
+  kernel_groups["check_symmetric"] = "check_gpu";
+  kernel_groups["check_diagonal_zeros"] = "check_gpu";
+
+  kernel_groups["dummy"] = "timing";
+
+  // Kernel group strings
+  // the dummy kernel is the only one not included in files
+  // so it is treated before the loop that iterates
+  // through  kernels to load all
+
+  kernel_strings["timing"] = dummy_kernel;
+  kernel_strings["check_gpu"] =
+#include <stan/math/prim/mat/kern_gpu/check_gpu.cl>  // NOLINT
+      ;                                              // NOLINT
+  kernel_strings["cholesky_decomposition"] =
+#include <stan/math/prim/mat/kern_gpu/cholesky_decomposition.cl>  // NOLINT
+      ;                                                           // NOLINT
+  kernel_strings["matrix_inverse"] =
+#include <stan/math/prim/mat/kern_gpu/matrix_inverse.cl>  // NOLINT
+      ;                                                   // NOLINT
+  kernel_strings["matrix_multiply"] =
+#include <stan/math/prim/mat/kern_gpu/matrix_multiply.cl>  // NOLINT
+      ;                                                    // NOLINT
+  kernel_strings["basic_matrix"] =
+#include <stan/math/prim/mat/kern_gpu/basic_matrix.cl>  // NOLINT
+      ;                                                 // NOLINT
+
+  // Check if the kernels were already compiled
+  compiled_kernels["basic_matrix"] = false;
+  compiled_kernels["matrix_multiply"] = false;
+  compiled_kernels["timing"] = false;
+  compiled_kernels["matrix_inverse"] = false;
+  compiled_kernels["cholesky_decomposition"] = false;
+  compiled_kernels["check_gpu"] = false;
+}
+
+// TODO(Rok): select some other platform/device than 0
+// TODO(Rok): option to turn profiling OFF
+/**
+ * The class that represents the OpenCL runtime
+ *
+ * The class is used to find the OpenCL supported
+ * platforms/devices and initialize the OpenCL
+ * context/command queue.
+ *
+ */
+class ocl {
+ private:
+  std::string description_;
+  cl::Context oclContext_;
+  cl::CommandQueue oclQueue_;
+  cl::Platform oclPlatform_;
+  cl::Device oclDevice_;
+
+  size_t max_workgroup_size;
+
+  inline void init() {
+    try {
+      std::vector<cl::Platform> allPlatforms;
+      cl::Platform::get(&allPlatforms);
+      if (allPlatforms.size() == 0) {
+        std::cout << " No platforms found. " << std::endl;
```

Comment:
This needs to go. We don't directly output to streams in our code any more.

Reply:
done